### PR TITLE
MIDI-24: Train 1D UNet to do peak detection

### PR DESCRIPTION
**Pull request by JasiekKaczmarczyk**
**[Train 1D UNet to do peak detection](https://sacrebleu.atlassian.net/browse/MIDI-24)**
Implementation of training loop. Here is link to wandb run: https://wandb.ai/jasiekkaczmarczyk/ecg-segmentation-unet/runs/bw1u2ttl/workspace?workspace=user-jasiekkaczmarczyk

To run training: 
```sh
python train.py hydra.job.chdir=False --config-name <config>
```

I tried overfitting single batch of size 16, which was successful, getting loss = 0.001309, with almost 100% acurracy and f1-score:
Zaznaczenie_003

Here is plot for validation set (also one batch), you can clearly see on loss chart that there is overfitting:
Zaznaczenie_004
It's a little bit weird that I got 98% accuracy and 93% f1-score on validation set.

I tried using onnx for storing trained model and uploaded it to wandb, here is model visualization: https://wandb.ai/jasiekkaczmarczyk/ecg-segmentation-unet/runs/bw1u2ttl/files/onnx-models/model-overfit-single-batch-2023-08-09-15-41.onnx

Next, I will train model on the whole dataset.

### CODE REVIEW

File: configs/config-default.yaml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  dataset_name: "roszcz/ecg-segmentation-ltafdb" # huggingface dataset
+  batch_size: 256
+  num_workers: 8
+  lr: 3e-4
+  weight_decay: 0.01
+  num_epochs: 10
+  accelerator: "gpu"
+  precision: 16
+  overfit_batches: 0
+  dataset_split_seed: 0 # seed to make random_split deterministic
```
Comment (roszcz) [inline configs/config-default.yaml]:
```suggestion
train:
  dataset_name: "roszcz/ecg-segmentation-ltafdb" # huggingface dataset
  batch_size: 256
  num_workers: 8
  lr: 3e-4
  weight_decay: 0.01
  num_epochs: 10
  accelerator: "gpu"
  precision: 16
  overfit_batches: 0
  dataset_split_seed: 0 # seed to make random_split deterministic
```

I think `hyperparameters` is not descriptive enough here and `train` seem like a better config section name. 

Notice that the `unet` params are also technically hyperparameters, but if you include that section here almost every config position would be in that section :)

**Comment by JasiekKaczmarczyk** [comment]
I created rough script to test model, and it's pretty surprising but this model overfitted for single batch actually can find peaks in a sample from validation set. Not perfectly, but has rough idea where they should be.
model_trained_single_batch_val_set

File: train.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+
+def preprocess_dataset(dataset_name: str, batch_size: int, num_workers: int, seed: int = 0):
+    dataset = ECGDataset(dataset_name)
+
+    # generator used to determine which records belong to which split
+    generator = torch.Generator().manual_seed(seed)
+
+    # spliting data based on seed
+    train_ds, val_ds, test_ds = random_split(dataset, [0.8, 0.1, 0.1], generator=generator)
```
Comment (roszcz) [inline train.py]: I think this will mix data from all the records together - my idea was to have the split defined by `record_id`, so something like this:

```python
ltafdb_records= ["100", ...]
train_records, test_records = random_split(ltafdb_records)
train_ds = Dataset(train_records)
test_ds = Dataset(test_records)
```

Let me know if this makes sense. Splitting the existing HF dataset by record id may be tricky, so maybe it would be better to create splits before pushing to hub

File: ecg_segmentation_dataset.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        # wrap signal and mask to torch.Tensor
+        signal = torch.tensor(record["signal"], dtype=torch.float32)
+        mask = torch.tensor(record["mask"], dtype=torch.float32)
+
+        # reshape data: [sequence_length, num_channels] -> [num_channels, sequence_length]
+        signal = einops.rearrange(signal, "l c -> c l")
+        # loss
+        mask = einops.rearrange(mask, "l c -> c l")
+
+        return signal, mask
```
Comment (roszcz) [inline ecg_segmentation_dataset.py]:
```suggestion
        record = self.data[index]

        # wrap signal and mask to torch.Tensor
        signal = torch.tensor(record["signal"], dtype=torch.float32)
        mask = torch.tensor(record["mask"], dtype=torch.float32)

        # reshape data: [sequence_length, num_channels] -> [num_channels, sequence_length]
        signal = einops.rearrange(signal, "l c -> c l")
        # loss
        mask = einops.rearrange(mask, "l c -> c l")

        item = {
          "signal": signal,
          "mask": mask,
          "record_id": record["record_id"],
        }
        return item
```

We may need the record id for better validation strategies. You'd also need to change the way batches are unpacked to something like this:

```python
for batch in loader:
  signal = batch["signal"]
  mask = batch["mask"]
```

File: train.py (hunk 2 above)
Comment (JasiekKaczmarczyk) [reply train.py]: Yeah, probably it was better to do them before pushing to hub

**Comment by roszcz** [comment]
@JasiekKaczmarczyk Is it possible that the one training batch included samples from the same exam that the one you used to make the plot? If that's the case this could still be a sign of overfitting 🤔

**Comment by roszcz** [comment]
I'm also not convinced if pytorch lightning interprets `overfit_batches=1` as "one batch", or as a "100% batches"? :)

**Comment by JasiekKaczmarczyk** [comment]
> I'm also not convinced if pytorch lightning interprets `overfit_batches=1` as "one batch", or as a "100% batches"? :)

It's one batch if you pass 1 as integer, 0 is all data, and if you pass float between 0-1 you get fraction of data. Also it wouldn't train that fast if it was whole dataset.

**Comment by roszcz** [comment]
I was worried about the confusion between `1` and `1.0` - looks light PL interprets those values as completely different. But I guess that would be easy to spot during training

**Comment by roszcz** [comment]
What's the loss value that you're getting with this? Testing your checkpoint locally I'm seeing values ~0.6, and as far as I can tell the model is learning some kind of very naive strategy.

Here are some samples I looked at:

image

image

It's easy to get high accuracy by guessing "mostly zero" (I'm getting ~88% with this model), but the f1 I get locally on the validation split is ~10%, which seems closer to what I'm seeing on the charts. 

On the train set accuracy I'm seeing is also below 90%, but f1 is higher. So I don't know if maybe I'm not loading the model correctly :thinking: 

Let me know if this makes sense, and please include your evaluation code in this repo, so we can be sure we're not discussing different things :)

**Comment by JasiekKaczmarczyk** [comment]
[Torch training loop](https://sacrebleu.atlassian.net/browse/MIDI-30)

I created second training loop and trained model for 1 epoch. It's working pretty well. I think the previous attempt failed because there was some bug in model checkpointing. Now, it's ok.

Here is result on test set:
model_trained_test_set

Here are metrics for each plotted sample:
Sample 0 acc: 0.9829999804496765, f1: 0.9119170904159546, iou: 0.8380952477455139
Sample 1 acc: 0.9959999918937683, f1: 0.9860140085220337, iou: 0.9724137783050537
Sample 2 acc: 0.9950000047683716, f1: 0.983818769454956, iou: 0.9681528806686401
Sample 3 acc: 0.9940000176429749, f1: 0.9790209531784058, iou: 0.9589040875434875

I also added some test scripts to the repo.

File: train.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        torch.save({"model": model.state_dict(), "optimizer": optimizer.state_dict()}, save_path)
+
+def step(
+        model: nn.Module,
+        batch: dict[str, torch.Tensor, torch.Tensor],
+        device: torch.device
+    ):
+    # extract signal and mask from batch
+        signal = batch["signal"].to(device)
+        mask = batch["mask"].to(device)
```
Comment (roszcz) [inline train.py]:
```suggestion
def save_checkpoint(model: nn.Module, optimizer: optim.Optimizer, save_path: str):
      # saving models
      torch.save({"model": model.state_dict(), "optimizer": optimizer.state_dict()}, save_path)

def step(
    model: nn.Module,
    batch: dict[str, torch.Tensor, torch.Tensor],
    device: torch.device
):
    # extract signal and mask from batch
    signal = batch["signal"].to(device)
    mask = batch["mask"].to(device)
```

Both those functions have indent issues - do you now how did this get pass pre commit hooks? 🤔

**Review by roszcz — APPROVED** [review]
Thanks @JasiekKaczmarczyk, this works great 👍 

Can you add a section to the README showing how to run this and what to expect for the results?
